### PR TITLE
Fit vectorizer using just training data in grant_tagger.py

### PR DESCRIPTION
**Pull request by lizgzil**
Fixing https://github.com/wellcometrust/nutrition-labels/issues/72

- The 2 functions for fitting the vectorizer (`fit_transform`) and the classifier (`fit`) and now condensed into one (`fit`)
- Uses a pipeline for the case where a vectorizer and a scaler need to be fit - no longer need to save out both separately
- `split_data` refactored to deal with a dataframe input of raw data rather than the vectorized data input
- update tests
- update latest results in the document
- rerun ensemble experiments with these models
- rerun ensemble with best parameters
- add a new ensemble config

Note: The best "ensemble" model after the experiments was actually not an ensemble (just bert+log reg). I've kept the pipeline as it is though so it could be used for multiple models or single models. However in actually using this model to make future predictions we can just used the GrantTagger class rather than the EnsembleGrantTagger class.

Best ensemble results: 0.9 for recall and precision THANK GOD 🎉

### CODE REVIEW

File: docs/Tech_Grants_2021.md [hunk 1]
```diff
@@ -101,29 +101,30 @@ A new model training config was made for training models (`configs/train_model/2
...
-| 210401 | bert       | naive_bayes | 0.753 | 0.64            | 0.912        | 0.007         | 0           ||
-| 210401 | bert       | SVM         | 0.848 | 0.824           | 0.875        | 0.642         | 0.457       ||
-| 210401 | bert       | log_reg     | 0.894 | 0.889           | 0.9          | 0.628         | 0.4         | x |
-| 210401 | scibert    | naive_bayes | 0.811 | 0.73            | 0.912        | 0             | 0           ||
-| 210401 | scibert    | SVM         | 0.835 | 0.789           | 0.888        | 0.75          | 0.514       ||
-| 210401 | scibert    | log_reg     | 0.861 | 0.835           | 0.888        | 0.689         | 0.5         | x |
+| Date   | Vectorizer | Classifier  | f1    | precision_score | recall_score | EPMC accuracy | RF accuracy |
+|--------|------------|-------------|-------|-----------------|--------------|---------------|-------------|
+| 210402 | count      | naive_bayes | 0.828 | 0.726           | 0.962        | 0.784         | 0.614       |
+| 210402 | count      | SVM         | 0.821 | 0.842           | 0.8          | 0.568         | 0.386       |
```
Comment (lizgzil) [inline docs/Tech_Grants_2021.md:116]: change from 210401 : better recall at expense of precision

File: docs/Tech_Grants_2021.md [hunk 2]
```diff
@@ -101,29 +101,30 @@ A new model training config was made for training models (`configs/train_model/2
...
+| 210402 | count      | log_reg     | 0.825 | 0.825           | 0.825        | 0.588         | 0.371       |
+| 210402 | tfidf      | naive_bayes | 0.811 | 0.7             | 0.962        | 0.811         | 0.657       |
+| 210402 | tfidf      | SVM         | 0.829 | 0.81            | 0.85         | 0.662         | 0.429       |
```
Comment (lizgzil) [inline docs/Tech_Grants_2021.md:119]: change from 210401 : better recall at expense of precision

**Comment by lizgzil** [comment]
Summary of the models that changed:

| Date   | Vectorizer | Classifier  | f1    | precision_score | recall_score | EPMC accuracy | RF accuracy | 
|--------|------------|-------------|-------|-----------------|--------------|---------------|-------------|
| 210401 | count      | SVM         | 0.816 | 0.896           | 0.75         | 0.568         | 0.386       |
| 210402 | count      | SVM         | 0.821 | 0.842           | 0.8          | 0.568         | 0.386       |
| 210401 | tfidf      | SVM         | 0.828 | 0.923           | 0.75         | 0.649         | 0.414       |
| 210402 | tfidf      | SVM         | 0.829 | 0.81            | 0.85         | 0.662         | 0.429       |
| 210401 | tfidf      | log_reg     | 0.824 | 0.863           | 0.788        | 0.709         | 0.457       |
| 210402 | tfidf      | log_reg     | 0.824 | 0.778           | 0.875        | 0.736         | 0.457       |
| 210401 | bert       | naive_bayes | 0.753 | 0.64            | 0.912        | 0.007         | 0           |
| 210402 | bert       | naive_bayes | 0.764 | 0.658           | 0.912        | 0.838         | 0.614       |
| 210401 | scibert    | naive_bayes | 0.811 | 0.73            | 0.912        | 0             | 0           |
| 210402 | scibert    | naive_bayes | 0.809 | 0.735           | 0.9          | 0.791         | 0.643       |
| 210401 | scibert    | log_reg     | 0.861 | 0.835           | 0.888        | 0.689         | 0.5         |
| 210402 | scibert    | log_reg     | 0.861 | 0.835           | 0.888        | 0.689         | 0.5         |

File: docs/Tech_Grants_2021.md [hunk 3]
```diff
@@ -101,29 +101,30 @@ A new model training config was made for training models (`configs/train_model/2
...
+| 210402 | tfidf      | log_reg     | 0.824 | 0.778           | 0.875        | 0.736         | 0.457       |
+| 210402 | bert       | naive_bayes | 0.764 | 0.658           | 0.912        | 0.838         | 0.614       |
```
Comment (nsorros) [inline docs/Tech_Grants_2021.md:121]: you did not have to rerun bert in theory since the vect fit does not affect them. why epmc and rf accuracy changed?

File: nutrition_labels/grant_tagger.py [hunk 4]
```diff
@@ -80,21 +81,19 @@ class GrantTagger:
...
-    fit_transform(data, train_data_id)
-        Fit a vectorizer using the text data
     transform(data)
         Vectorize text data using the fitted vectorizer
-    split_data(X_vect, y)
+    split_data(data, train_data_id)
         Split the data into train and test sets
     fit(X, y)
-        Train the model
+        Fit a vectorizer using the text data, and train the model
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:89]: I would also agree with the previous text, the model being the combination of vectorizer and classifier

File: nutrition_labels/grant_tagger.py [hunk 5]
```diff
@@ -216,20 +179,59 @@ def split_data(self, X_vect, y):
...
+                token_pattern=r"(?u)\b\w+\b",
+                ngram_range=(1, 2),
+                stop_words='english'
+            )
+        elif self.vectorizer_type == "tfidf":
+            self.vectorizer = TfidfVectorizer(
+                analyzer="word",
+                token_pattern=r"(?u)\b\w+\b",
+                ngram_range=(1, 2),
+                stop_words='english'
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:226]: I think I might have mentioned that in the past but these params seem set in stone, it would be nice to be able to pass different params from the config.

File: nutrition_labels/grant_tagger.py [hunk 6]
```diff
@@ -290,29 +297,28 @@ def train_test_info(self):
         }
         """
 
-        y_predict = self.model.predict(self.X_vect)
-        y_predict_proba = self.model.predict_proba(self.X_vect)
-        # The probability of the predicted binary value
-        y_predict_proba = np.max(y_predict_proba, axis=1)
+        X_test_vect = self.vectorizer.transform(X_test)
+        X_unused_vect = self.vectorizer.transform(self.X_unused)
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:301]: `X_unused` is not passed in, where did the class find it? can it be the case that this method is passed without `X_unused` being defined? also X_unused sounds like a heavy object to keep in memory and save with the model, is it necessary?

File: nutrition_labels/grant_tagger.py [hunk 7]
```diff
@@ -324,9 +330,6 @@ def save_model(self, output_path, evaluation_results=None):
             pickle.dump(self.model, f)
         with open(os.path.join(output_path, "vectorizer.pickle"), "wb") as f:
             pickle.dump(self.vectorizer, f)
-        if self.scaler:
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:327]: not that you can pickle the entire pipeline, you don't need to pickle components separately

File: tests/test_grant_tagger.py [hunk 8]
```diff
@@ -137,13 +130,30 @@ def test_train_test_info():
         prediction_cols=prediction_cols,
         label_name=label_name,
         )
-
-    X_vect, y = grant_tagger.fit_transform(training_data, train_data_id)
-    X_train, X_test, y_train, y_test = grant_tagger.split_data(X_vect, y)
+    X_train, X_test, y_train, y_test = grant_tagger.split_data(training_data, train_data_id)
```
Comment (nsorros) [inline tests/test_grant_tagger.py:133]: what is the point of having a split_data function in the class? is this function the responsibility of the class? also what does it promise to the user?

split_data sounds like a function that should live outside the class

File: nutrition_labels/grant_tagger.py [hunk 9]
```diff
@@ -216,20 +179,59 @@ def split_data(self, X_vect, y):
...
+        y = data[self.label_name].tolist()
 
         # Randomly shuffled at the beginning so turn this off
         X_train, X_test, y_train, y_test = train_test_split(
-            X_vect, y, test_size=self.test_size, shuffle=False
-        )
+                X, y, test_size=self.test_size, shuffle=False
+            )
+
+        self.X_train_ids = X_train[train_data_id].tolist()
```
Comment (nsorros) [inline nutrition_labels/grant_tagger.py:204]: why do you save that? note that if you rely on the user keeping the data in the same order so that ids maps to training examples, I think this can lead to bugs. a user can receive X_train and change the order by shuffling.

**Review by nsorros — COMMENTED** [review]
I think there are some points that might lead to bugs so leaving this open for a discussion before I approve

File: docs/Tech_Grants_2021.md (hunk 3 above)
Comment (lizgzil) [reply docs/Tech_Grants_2021.md:121]: True, but it was a nice sanity check to see that they didn't all change wildly! The BERT/SciBERT+naive bayes models did change though - but this is because in those cases I needed the minmaxscaler for these - and since this was previously fitted using all the data, I guess the scaling is very slightly different when fitted using just the training data (reflected in very small changes to the overall metrics).

The EPMC and RF accuracy changed because I had forgotten to update them after I included the minmaxscaler in the pipeline - so the 210401 result for this was wrong anyway.

File: nutrition_labels/grant_tagger.py (hunk 5 above)
Comment (lizgzil) [reply nutrition_labels/grant_tagger.py:226]: I agree, I'll add this as an issue rather than fix here https://github.com/wellcometrust/nutrition-labels/issues/74

File: nutrition_labels/grant_tagger.py (hunk 6 above)
Comment (lizgzil) [reply nutrition_labels/grant_tagger.py:301]: it is created when X_test and X_train are created in `split_data`. `train_test_info` is only really designed to be useful to run after training (and therefore `self.X_unused` should exist). But I see your point and maybe `split_data` should return `return X_train, X_test, X_unused, y_train, y_test, y_unused` - a bit unfortunate that it's not in the style of sklearn split though.

File: nutrition_labels/grant_tagger.py (hunk 9 above)
Comment (lizgzil) [reply nutrition_labels/grant_tagger.py:204]: I save it to be able to link IDs to whether they were in the test/train/unseen split, since this was something I wanted for my pipeline.

If you're pipeline is (as mine is in `train_several_models`):
```
X_train, X_test, y_train, y_test = grant_tagger.split_data(training_data, train_data_id)
grant_tagger.fit(X_train, y_train)
grant_info = grant_tagger.train_test_info(X_test, y_train, y_test)
```
then you will save out a nice file of `IDs - actual Y - predicted Y - whether this ID is in the train/test/unseen split`.

You are right that if you did
```
X_train, X_test, y_train, y_test = grant_tagger.split_data(training_data, train_data_id)
grant_tagger.fit(X_train, y_train)
X_test = X_test[shuffle]
y_train = y_train[shuffle]
y_test = y_test[shuffle]
grant_info = grant_tagger.train_test_info(X_test, y_train, y_test)
```
you could mess up this output file (if indeed the user even wanted to output this file).

File: tests/test_grant_tagger.py (hunk 8 above)
Comment (lizgzil) [reply tests/test_grant_tagger.py:133]: It's useful to keep it in the class to keep a record of which IDs were in the train/test/unseen split. But its true this function could return everything, i.e.
```
def split_data(data, train_data_id="Internal ID", label_name='Relevance code', split_seed=42, relevant_sample_ratio=1):
...
return X_train, X_test, X_unused, y_train, y_test, y_unused, X_train_ids, X_test_ids, unused_ids
```

**Comment by lizgzil** [comment]
@nsorros I've addressed your comments in https://github.com/wellcometrust/nutrition-labels/pull/73/commits/8311183dc3876cf4590330987648f2af06328fc5 - I've made it so that the ids aren't stored in class variables, so it should be clearer.
I chose to keep `split_data` in the class since it's very much tied to the structure of all the functions for this project, and since it's not a bug and time is of the essence, it didn't seem essential to change this at this point.

File: nutrition_labels/grant_tagger.py (hunk 6 above)
Comment (nsorros) [reply nutrition_labels/grant_tagger.py]: I think that would be better and there is no way to conform to the sklearn api for all their functions.

File: tests/test_grant_tagger.py (hunk 8 above)
Comment (nsorros) [reply tests/test_grant_tagger.py]: note that this can be a source of problems as I mentioned in that it returns a list of data and keeps an order with the ids, hoping that the user will not shuffle or split in any way

**Review by nsorros — APPROVED** [review]
